pkg/templates/kubeadm/v1beta4/kubeadm_test.go

```diff
@@ -0,0 +1,314 @@
+		},
+	}
+	expectedAPIServerExtraArgs := []kubeadmv1beta4.Arg{
+		{
+			Name:  "existing-flag",
+			Value: "existing-value",
+		},
+		{
+			Name:  "flag5",
+			Value: "value5",
+		},
+		{
+			Name:  "flag6",
+			Value: "value6",
+		},
+		{
+			Name:  "feature-gates",
+			Value: "feature5=true,feature6=false",
+		},
+	}
+	expectedControllerManagerFeatureGates := "feature1=true,feature2=false"
+	expectedSchedulerFeatureGates := "feature3=true,feature4=false"
+	expectedAPIServerFeatureGates := "feature5=true,feature6=false"
+
+	addControlPlaneComponentsAdditionalArgs(cluster, clusterConfig)
+
+	// Check ControllerManager
+	if !areArgsEqual(clusterConfig.ControllerManager.ExtraArgs, expectedControllerManagerExtraArgs) {
+		t.Errorf("ControllerManager ExtraArgs mismatch, got: %v, want: %v", clusterConfig.ControllerManager.ExtraArgs, expectedControllerManagerExtraArgs)
+	}
+	fgs, _ := kubeadmv1beta4.GetArgValue(clusterConfig.ControllerManager.ExtraArgs, "feature-gates", -1)
+	if !areFeatureGatesEqual(fgs, expectedControllerManagerFeatureGates) {
+		t.Errorf("ControllerManager FeatureGates mismatch, got: %s, want: %s", fgs, expectedControllerManagerFeatureGates)
+	}
+
+	// Check Scheduler
+	if !areArgsEqual(clusterConfig.Scheduler.ExtraArgs, expectedSchedulerExtraArgs) {
+		t.Errorf("Scheduler ExtraArgs mismatch, got: %v, want: %v", clusterConfig.Scheduler.ExtraArgs, expectedSchedulerExtraArgs)
+	}
+	fgs, _ = kubeadmv1beta4.GetArgValue(clusterConfig.Scheduler.ExtraArgs, "feature-gates", -1)
+	if !areFeatureGatesEqual(fgs, expectedSchedulerFeatureGates) {
+		t.Errorf("Scheduler FeatureGates mismatch, got: %s, want: %s", fgs, expectedSchedulerFeatureGates)
+	}
+
+	// Check APIServer
+	if !areArgsEqual(clusterConfig.APIServer.ExtraArgs, expectedAPIServerExtraArgs) {
+		t.Errorf("APIServer ExtraArgs mismatch, got: %v, want: %v", clusterConfig.APIServer.ExtraArgs, expectedAPIServerExtraArgs)
+	}
+	fgs, _ = kubeadmv1beta4.GetArgValue(clusterConfig.APIServer.ExtraArgs, "feature-gates", -1)
+	if !areFeatureGatesEqual(fgs, expectedAPIServerFeatureGates) {
+		t.Errorf("APIServer FeatureGates mismatch, got: %s, want: %s", fgs, expectedAPIServerFeatureGates)
+	}
+}
+
+// areFeatureGatesEqual compares two feature gates strings irrespective of the order of the feature gates.
+func areFeatureGatesEqual(result, expected string) bool {
+	resultMap := make(map[string]bool)
+	expectedMap := make(map[string]bool)
+
+	// Parse the result string into a map
+	for _, fg := range strings.Split(result, ",") {
+		kv := strings.Split(fg, "=")
+		if len(kv) == 2 {
+			resultMap[kv[0]] = kv[1] == "true"
+		}
+	}
+
+	// Parse the expected string into a map
+	for _, fg := range strings.Split(expected, ",") {
+		kv := strings.Split(fg, "=")
+		if len(kv) == 2 {
+			expectedMap[kv[0]] = kv[1] == "true"
+		}
+	}
+
+	// Compare the maps
+	if len(resultMap) != len(expectedMap) {
+		return false
+	}
+
+	for k, v := range resultMap {
+		if expectedMap[k] != v {
```

Comment:
Could this fail because of the implicit zero value? What if `resultMap = {"foo":false}` and `expectedMap = {"bla":true]`, then it would fail for k = `foo`, no?